sycl/include/sycl/types.hpp

```diff
@@ -580,6 +553,41 @@ template <typename Type, int NumElements> class vec {
   // in the class, so vec<float, 16> should be equal to float16 in memory.
   using DataType = typename detail::VecStorage<DataT, NumElements>::DataType;
 
+  // This represents HOW  we will approach the underlying value, so as to
+  // benefit from vector speed improvements
+  using VectorDataType =
+      typename detail::VecStorage<DataT, NumElements>::VectorDataType;
+
+  VectorDataType &getAsVector() {
+    return *reinterpret_cast<VectorDataType *>(m_Data.data());
```

Comment:
We should leave a comment here that breaking strict aliasing rules is a UB. Doing this could lead to weird issues depending on how optimizations work

Reply:
If the size is below 64 bytes, strict aliasing is the only UB problem. Is my understanding correct that all our supported compilers (MSVC, gcc, clang) don't produce incorrect results for this (because while UB it is too common in real code)? 

Above 64 bytes this is also UB because the alignment is wrong. Can this produce incorrect results?

Reply:
@rolandschulz  the max alignment is 64 (as was agreed in the JIRA discussion) and that's what I've implemented here, and in that case we switch to std::array instead of ext_vector_type. So, no incorrect results on that front.

Reply:
The max alignment of 64 is for the storage. The storage is  using `DataType` which is `ext_vector_type` up to 64 bytes,  and array with 64 byte alignment otherwise. But the return type is `VectorDataType`. And `VectorDataType` is always `ext_vector_type` (if supported by compiler). Is all this correct? 
If so the problem is that you are returning a reference with a type with >64 byte alignment which refers to storage with 64 byte alignment. In the JIRA we didn't discuss what we do with the return type of getAsVector.
I don't see how we can implement getAsVector with the spec as-is.

It would be possible if we were to change the spec to say:
- `getAsVector` is only available up to 64 bytes, or
- return by value rather than reference, or
- return by value above 64 bytes

@gmlueck Do you think one of the changes would be acceptable?

Reply:
Sorry I confused `getAsVector` with `operator vector_t()`. Only the later is in the spec. And it returns it by value. Therefore I agree this isn't a spec concern. 
We still need to make sure that the internal usage gives correct results for >64bytes.

Reply:
I only see one usage of getAsVector (impl of binop). If we replace that with bitcast we don't have either UB problem.